### ISSUE TITLE
Promoted transactions in TxPool not demoting on recoverable error

### ISSUE THREAD

**Issue by lazartravica** [0xPolygon/polygon-edge#731]
Was able to repro what we believe to be a similar scenario to the one we initially observed using the steps below:

1. assuming account balance is 5 ETH - submit a couple simple transfer txs we can currently afford with current nonce + 10 (for example each tx costing total of 3 ETH each), these txs will then appear within `queued` in txpool_inspect response (for example with nonces 450-451)
2. submit 10 txs we can currently afford using the current nonce (for example nonces 440-449 with each tx costing only 0.1 ETH each)
3. once 449 has been queued, promoted and mined, the queued txs with nonces 450 and 451 will be promoted to pending however at this point, the account balance cannot cover the tx costs:
```
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {
            "0xFd9E636956415cC447Fb621223d2fc779819E533": {
                "450": "3000000000000000000 wei + 21000 gas x 10000000000 wei",
                "451": "3000000000000000000 wei + 21000 gas x 10000000000 wei"
            }
        },
        "queued": {},
        "currentCapacity": 2,
        "maxCapacity": 4096
    }
}
```
4. these stuck pending txs do not go away on their own even once the account balance is topped up
---

Not sure but I would expect the queued txs to consider the account balance at the latest block during the promotion process and therefore not get promoted.

_Originally posted by @dankostiuk in https://github.com/0xPolygon/polygon-edge/issues/711#issuecomment-1244680548_

**Comment by dbrajovic** [comment]
Possibly related:

https://github.com/0xPolygon/polygon-edge/issues/226

**Comment by dbrajovic** [comment]
@dankostiuk 

This looks a lot like what you previously reported back in #226. The part which raises concern is:

> these stuck pending txs do not go away on their own 

even more so

> even once the account balance is topped up

The `Demote -> Drop` logic was implemented to solve this specific problem. Continuing from the previous issue, how come there is no `Demote/Drop` taking place for pending `450` and `451`? The only way I see that _not_ happening is that the txs eventually ended up in a block (due to balances being topped), but you claim this is not the case.

If you have the logs (or can reproduce them again), would you mind sharing?

**Comment by dankostiuk** [comment]
Hey @dbrajovic ,

I've reproduced the issue with an account with 101 ETH and current nonce (getTransactionCount) 566. See below for the steps I took:

1. Submit 2 txs at nonce 576 and 577 (each 100 ETH transfers) 
```
{"message":"2022-09-13T17:26:42.677Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x3f9feffb5d9f2fe43aec2f6d1762a7e8bf65055ffa56cd3ed08c8346b786731c"}
{"message":"2022-09-13T17:26:42.678Z [DEBUG] polygon.txpool: enqueue request: hash=0x3f9feffb5d9f2fe43aec2f6d1762a7e8bf65055ffa56cd3ed08c8346b786731c"}
{"message":"2022-09-13T17:26:42.678Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x3f9feffb5d9f2fe43aec2f6d1762a7e8bf65055ffa56cd3ed08c8346b786731c"}
{"message":"2022-09-13T17:26:42.679Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x3f9feffb5d9f2fe43aec2f6d1762a7e8bf65055ffa56cd3ed08c8346b786731c"}
{"message":"2022-09-13T17:25:44.757Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x622f8fcad7fa92ec444c7c0154625be7ae657315b397052afcc9c652a42b7172"}
{"message":"2022-09-13T17:25:44.757Z [ERROR] polygon.txpool: failed to add tx: err=\"cannot extract signature\""}
{"message":"2022-09-13T17:25:44.687Z [DEBUG] polygon.txpool: add tx: origin=local hash=0xc23331c796574bd40bc6605a4682798268e1b44d27c8c16ea044510cf8319362"}
{"message":"2022-09-13T17:25:44.687Z [DEBUG] polygon.txpool: enqueue request: hash=0xc23331c796574bd40bc6605a4682798268e1b44d27c8c16ea044510cf8319362"}
{"message":"2022-09-13T17:25:44.688Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0xc23331c796574bd40bc6605a4682798268e1b44d27c8c16ea044510cf8319362"}
{"message":"2022-09-13T17:25:44.688Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0xc23331c796574bd40bc6605a4682798268e1b44d27c8c16ea044510cf8319362"}
```
2. Observe txpool_inspect:
```
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {},
        "queued": {
            "0xFd9E636956415cC447Fb621223d2fc779819E533": {
                "576": "100000000000000000000 wei + 21000 gas x 1000000000000 wei",
                "577": "100000000000000000000 wei + 21000 gas x 1000000000000 wei"
            }
        },
        "currentCapacity": 2,
        "maxCapacity": 4096
    }
}
```
3. Submit 9 txs of 1 ETH transfers at nonces 566-575 bringing the account balance down to 91 ETH:
```
{"message":"2022-09-13T17:34:11.967Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x740b7078bebd106441143363bbeece5d39499f26d0a25c447553819015815067"}
{"message":"2022-09-13T17:34:11.968Z [DEBUG] polygon.txpool: enqueue request: hash=0x740b7078bebd106441143363bbeece5d39499f26d0a25c447553819015815067"}
{"message":"2022-09-13T17:34:11.968Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{574 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1329 30923941666030945867279580372060681180842287819115193999198565500534724160735 21656415348268990682279983027931692126914737405218610315352285877768957231060 0x740b7078bebd106441143363bbeece5d39499f26d0a25c447553819015815067 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.968Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x740b7078bebd106441143363bbeece5d39499f26d0a25c447553819015815067"}
{"message":"2022-09-13T17:34:11.968Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x740b7078bebd106441143363bbeece5d39499f26d0a25c447553819015815067"}
{"message":"2022-09-13T17:34:11.842Z [DEBUG] polygon.txpool: enqueue request: hash=0x6c6e91aa3e2a4430bdfd76e5d297ccc9f8b73d1d6900886f5e66865ba19100d6"}
{"message":"2022-09-13T17:34:11.842Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{572 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1330 40033251082726511871560053225038380089825559496342878620455674423157950751854 40194805603506121168763013449274507676296678975547380201648772564519379967843 0x6c6e91aa3e2a4430bdfd76e5d297ccc9f8b73d1d6900886f5e66865ba19100d6 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.842Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x6c6e91aa3e2a4430bdfd76e5d297ccc9f8b73d1d6900886f5e66865ba19100d6"}
{"message":"2022-09-13T17:34:11.842Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x6c6e91aa3e2a4430bdfd76e5d297ccc9f8b73d1d6900886f5e66865ba19100d6"}
{"message":"2022-09-13T17:34:11.907Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x23fc715f7080bbe8a9e293b26ba583aea10c02274670aba0d6827013bb64095b"}
{"message":"2022-09-13T17:34:11.907Z [DEBUG] polygon.txpool: enqueue request: hash=0x23fc715f7080bbe8a9e293b26ba583aea10c02274670aba0d6827013bb64095b"}
{"message":"2022-09-13T17:34:11.907Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{573 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1330 93923435324653423583825012471354292344587904735953679887037846553974817293975 26638792935360259809425414208516810442505162842805971209354047471436087385014 0x23fc715f7080bbe8a9e293b26ba583aea10c02274670aba0d6827013bb64095b 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.908Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x23fc715f7080bbe8a9e293b26ba583aea10c02274670aba0d6827013bb64095b"}
{"message":"2022-09-13T17:34:11.908Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x23fc715f7080bbe8a9e293b26ba583aea10c02274670aba0d6827013bb64095b"}
{"message":"2022-09-13T17:34:11.781Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x54b169b28cf8ae5a2be73206da260a96548e802aebfd10f6da72910a45f6c412"}
{"message":"2022-09-13T17:34:11.841Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x6c6e91aa3e2a4430bdfd76e5d297ccc9f8b73d1d6900886f5e66865ba19100d6"}
{"message":"2022-09-13T17:34:11.648Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x55884add7fded14b03abce21117e5915315dafee92817d75eddad14798abab78"}
{"message":"2022-09-13T17:34:11.649Z [DEBUG] polygon.txpool: enqueue request: hash=0x55884add7fded14b03abce21117e5915315dafee92817d75eddad14798abab78"}
{"message":"2022-09-13T17:34:11.649Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{569 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1330 41859158353927946307637015435035345840300970918534051313067693390731877508116 45297566749234101650234167424860916276678800750123092333466256176852964367086 0x55884add7fded14b03abce21117e5915315dafee92817d75eddad14798abab78 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.649Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x55884add7fded14b03abce21117e5915315dafee92817d75eddad14798abab78"}
{"message":"2022-09-13T17:34:11.650Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x55884add7fded14b03abce21117e5915315dafee92817d75eddad14798abab78"}
{"message":"2022-09-13T17:34:11.713Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x6dd846da95b31decbc647db06121857d85b3edd9b2c60c4cae6216dc163cc940"}
{"message":"2022-09-13T17:34:11.713Z [DEBUG] polygon.txpool: enqueue request: hash=0x6dd846da95b31decbc647db06121857d85b3edd9b2c60c4cae6216dc163cc940"}
{"message":"2022-09-13T17:34:11.713Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{570 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1329 21562495940526794728475752856136169800716612180683079360302411951870085481030 11076613169394570796063111821306040817050300567328699278638525336855263086906 0x6dd846da95b31decbc647db06121857d85b3edd9b2c60c4cae6216dc163cc940 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.713Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x6dd846da95b31decbc647db06121857d85b3edd9b2c60c4cae6216dc163cc940"}
{"message":"2022-09-13T17:34:11.714Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x6dd846da95b31decbc647db06121857d85b3edd9b2c60c4cae6216dc163cc940"}
{"message":"2022-09-13T17:34:11.780Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x54b169b28cf8ae5a2be73206da260a96548e802aebfd10f6da72910a45f6c412"}
{"message":"2022-09-13T17:34:11.780Z [DEBUG] polygon.txpool: enqueue request: hash=0x54b169b28cf8ae5a2be73206da260a96548e802aebfd10f6da72910a45f6c412"}
{"message":"2022-09-13T17:34:11.780Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{571 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1330 31488620698489643013016275544947242038887968899512017763413052905551337877017 13226584650215278037543283922478597435611245266801153831408049697630248114196 0x54b169b28cf8ae5a2be73206da260a96548e802aebfd10f6da72910a45f6c412 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.780Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x54b169b28cf8ae5a2be73206da260a96548e802aebfd10f6da72910a45f6c412"}
{"message":"2022-09-13T17:34:11.577Z [DEBUG] polygon.txpool: add tx: origin=local hash=0xb5772ab00a1a2b1f5f0d003ce836f39496b132412d6c9d44dac1fc08dc8f1ffc"}
{"message":"2022-09-13T17:34:11.578Z [DEBUG] polygon.txpool: enqueue request: hash=0xb5772ab00a1a2b1f5f0d003ce836f39496b132412d6c9d44dac1fc08dc8f1ffc"}
{"message":"2022-09-13T17:34:11.578Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{568 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1330 53370140039377173349782426586430664268719835532064242247299972802667921285446 6508992719347488683148848650463735134696879087230639577146215447838373423785 0xb5772ab00a1a2b1f5f0d003ce836f39496b132412d6c9d44dac1fc08dc8f1ffc 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.578Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0xb5772ab00a1a2b1f5f0d003ce836f39496b132412d6c9d44dac1fc08dc8f1ffc"}
{"message":"2022-09-13T17:34:11.578Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0xb5772ab00a1a2b1f5f0d003ce836f39496b132412d6c9d44dac1fc08dc8f1ffc"}
{"message":"2022-09-13T17:34:11.506Z [DEBUG] polygon.txpool: add tx: origin=local hash=0xd1cd145b636875e08adf10f95f26104317a960b6900161023a7c2024fb40c609"}
{"message":"2022-09-13T17:34:11.507Z [DEBUG] polygon.txpool: enqueue request: hash=0xd1cd145b636875e08adf10f95f26104317a960b6900161023a7c2024fb40c609"}
{"message":"2022-09-13T17:34:11.507Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{567 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1329 92933699373834567443126389302639783639386082582048613664710554712876713077846 24666821731964205288432932041213178555905726942112892860008326175362054387906 0xd1cd145b636875e08adf10f95f26104317a960b6900161023a7c2024fb40c609 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:34:11.507Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0xd1cd145b636875e08adf10f95f26104317a960b6900161023a7c2024fb40c609"}
{"message":"2022-09-13T17:34:11.508Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0xd1cd145b636875e08adf10f95f26104317a960b6900161023a7c2024fb40c609"}
{"message":"2022-09-13T17:33:55.607Z [DEBUG] polygon.txpool: add tx: origin=local hash=0xf460a264ae8eeca01c2c103b90f7b791999dba673e5cc243677ada597f15da9d"}
{"message":"2022-09-13T17:33:55.608Z [DEBUG] polygon.txpool: enqueue request: hash=0xf460a264ae8eeca01c2c103b90f7b791999dba673e5cc243677ada597f15da9d"}
{"message":"2022-09-13T17:33:55.608Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{566 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 1000000000000000000 [] 1329 79719255073341504267607482991611943503326171675126109543866035325383102464129 36774335660565653461805963173286370310494612798350888554401389915823991039897 0xf460a264ae8eeca01c2c103b90f7b791999dba673e5cc243677ada597f15da9d 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:33:55.608Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0xf460a264ae8eeca01c2c103b90f7b791999dba673e5cc243677ada597f15da9d"}
{"message":"2022-09-13T17:33:55.609Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0xf460a264ae8eeca01c2c103b90f7b791999dba673e5cc243677ada597f15da9d"}
```
4. Submit 1 tx of 1 ETH transfer at nonce 575, you can see 575 gets promoted and cleared as the balance could afford the tx but the other 2 txs get stuck in pending as the account cannot afford either of them. 

txpool_inspect:
```
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {
            "0xFd9E636956415cC447Fb621223d2fc779819E533": {
                "576": "100000000000000000000 wei + 21000 gas x 1000000000000 wei",
                "577": "100000000000000000000 wei + 21000 gas x 1000000000000 wei"
            }
        },
        "queued": {},
        "currentCapacity": 2,
        "maxCapacity": 4096
    }
}
```
It does appear that the txs are demoted as the account gets dropped as can be seen in the logs below, but we suspect that something weird is going on possibly due to what's going on in https://github.com/0xPolygon/polygon-edge/issues/711 :
```
{"message":"2022-09-13T17:40:57.478Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:40:57.478Z [DEBUG] polygon.txpool: dropped account txs: num=2 next_nonce=576 address=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T17:40:06.897Z [ERROR] polygon: failed to apply tx: err=\"not enough funds for transfer with given value\""}
{"message":"2022-09-13T17:39:56.786Z [ERROR] polygon: failed to apply tx: err=\"not enough funds for transfer with given value\""}
{"message":"2022-09-13T17:39:46.669Z [ERROR] polygon: failed to apply tx: err=\"not enough funds for transfer with given value\""}
{"message":"2022-09-13T17:39:36.559Z [ERROR] polygon: failed to apply tx: err=\"not enough funds for transfer with given value\""}
{"message":"2022-09-13T17:39:26.444Z [ERROR] polygon: failed to apply tx: err=\"not enough funds for transfer with given value\""}
{"message":"2022-09-13T17:39:16.326Z [ERROR] polygon: failed to apply tx: err=\"not enough funds for transfer with given value\""}
```

**Comment by dankostiuk** [comment]
Strangely after running the same test again but on a different chain, I can confirm that the pending txs do not get demoted as expected, even after waiting for 30 minutes. The only way to clear it is to submit another transaction with the current nonce (469) or restarting the node..

txpool_content:
```
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {
            "0xFd9E636956415cC447Fb621223d2fc779819E533": {
                "469": {
                    "nonce": "0x1d5",
                    "gasPrice": "0xe8d4a51000",
                    "gas": "0x5208",
                    "to": "0x86135362aAAf5C00E51b003fb00A34c72A88027A",
                    "value": "0x3782dace9d900000",
                    "input": "0x",
                    "hash": "0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b",
                    "from": "0xFd9E636956415cC447Fb621223d2fc779819E533",
                    "blockHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
                    "blockNumber": null,
                    "transactionIndex": null
                }
            }
        },
        "queued": {}
    }
}
```

Strangely I see no `failed to apply tx` due to `not enough funds for transfer with given value` this time:
```
{"message":"2022-09-13T21:28:40.983Z [ERROR] polygon.txpool: failed to add tx: err=\"already known\""}
{"message":"2022-09-13T21:28:40.983Z [ERROR] polygon.dispatcher: failed to dispatch: method=eth_sendRawTransaction err=\"already known\""}
{"message":"2022-09-13T21:28:36.160Z [ERROR] polygon.txpool: failed to add tx: err=\"already known\""}
{"message":"2022-09-13T21:28:36.160Z [ERROR] polygon.dispatcher: failed to dispatch: method=eth_sendRawTransaction err=\"already known\""}
{"message":"2022-09-13T21:28:31.776Z [ERROR] polygon.txpool: failed to add tx: err=\"already known\""}
{"message":"2022-09-13T21:28:31.776Z [ERROR] polygon.dispatcher: failed to dispatch: method=eth_sendRawTransaction err=\"already known\""}
{"message":"2022-09-13T21:27:45.065Z [ERROR] polygon.txpool: failed to add tx: err=\"already known\""}
{"message":"2022-09-13T21:27:45.065Z [ERROR] polygon.dispatcher: failed to dispatch: method=eth_sendRawTransaction err=\"already known\""}
{"message":"2022-09-13T21:20:04.280Z [ERROR] polygon.txpool: failed to add tx: err=\"already known\""}
{"message":"2022-09-13T21:20:04.280Z [ERROR] polygon.dispatcher: failed to dispatch: method=eth_sendRawTransaction err=\"already known\""}
{"message":"2022-09-13T21:17:50.755Z [ERROR] polygon.txpool: failed to add tx: err=\"already known\""}
{"message":"2022-09-13T21:17:50.755Z [ERROR] polygon.dispatcher: failed to dispatch: method=eth_sendRawTransaction err=\"already known\""}
{"message":"2022-09-13T21:16:28.215Z [DEBUG] polygon.txpool: promote request: promoted=[\"&{468 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 100000000000000000 [] 868 30056756898382267316677801622757651564927143937259979382761353626455302926383 47113770464537078203437525260827126903118628152657375285955440147401237640795 0x7ad1e3fe65c3e2a8103edb7a39df8693eea1cf765ab0aedecf930759b0162425 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\", \"&{469 1000000000000 21000 0x86135362aAAf5C00E51b003fb00A34c72A88027A 4000000000000000000 [] 868 42697515984464300657278602548638824612254944262271935529900370135821422896885 57794567439952944520874537018797889559482448514199087137591755037331737829888 0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b 0xFd9E636956415cC447Fb621223d2fc779819E533 {114}}\"] addr=0xFd9E636956415cC447Fb621223d2fc779819E533"}
{"message":"2022-09-13T21:15:37.076Z [DEBUG] polygon.txpool: rejecting known tx (gossip): hash=0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b"}
{"message":"2022-09-13T21:15:37.075Z [DEBUG] polygon.txpool: add tx: origin=local hash=0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b"}
{"message":"2022-09-13T21:15:37.076Z [DEBUG] polygon.txpool: enqueue request: hash=0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b"}
{"message":"2022-09-13T21:15:37.076Z [DEBUG] polygon.jsonrpc: handle: response=\"{\\\"jsonrpc\\\":\\\"2.0\\\",\\\"id\\\":9,\\\"result\\\":\\\"0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b\\\"}\""}
{"message":"2022-09-13T21:15:37.076Z [DEBUG] polygon.txpool: add tx: origin=gossip hash=0x8d971270d968eef068e8ccff978eef68955418ebba3c41ef59fb9b48d54d471b"}
```

**Comment by dankostiuk** [comment]
Hey @dbrajovic,

After further investigation, it appears that the demote logic is only triggered after `buildProposal()` is hit, which only ever occurs if the node is a validator.

Our RPC cluster consists of solely non-validating nodes, so any invalid pending transactions within their txpools will never clear.

I suppose the solution could be to have validators gossip to non-validating nodes to replay the same demotion across all their pools.

**Comment by dbrajovic** [comment]
> Our RPC cluster consists of solely non-validating nodes

Ah, that explains a lot. 

In that case, we'll probably discuss a sort-of feature for non-validators specifically, so they're able to gossip transactions to validators in a way that would not block their txpool.

I'll keep this issue open until we push out the PR.

**Comment by mrwillis** [comment]
Maybe a nitpick, but although you guys have marked it as a feature, I would argue it's a bug still as it's very common to have RPC providers (non-validators) for chains. Effectively this means that if you're out of $, from the view of the user connected to an RPC node, your account is bricked until the txpool is somehow cleared, or you're smart enough to resubmit with the same nonce.

**Comment by Kourin1996** [comment]
@mrwillis @dankostiuk Hello, I've created a PR. It adds a function to clean up inactive accounts' transactions in non-validator's TxPool. It will be merged in the next MINOR or PATCH release.

https://github.com/0xPolygon/polygon-edge/pull/742

**Comment by Kourin1996** [comment]
The above PR is going to be included at 0.5.1 release
